### DISCUSSION TITLE
How to build a freestanding Zephyr application with cmake and venv?

### DISCUSSION THREAD

**Discussion by sebistark** [zephyrproject-rtos/zephyr#75395] [Q&A] [answered]
# How to build a freestanding Zephyr application with cmake and venv?
And how does it work when python modules were installed within python venv?

I always get an error `ModuleNotFoundError: No module named 'pykwalify'`.

the global pip list does not contain this module, but the pip list in venv contains it, as expected.

Here is the whole log:

```
**** CMake Configure of configuration build for project demoapp ****
/usr/bin/cmake -S . -B build -GUnix Makefiles 
CMake Warning:
  Ignoring empty string ("") provided on the command line.


Loading Zephyr default modules (Freestanding).
-- Application: /home/xxx/workspace_STM/demoapp
-- CMake version: 3.29.6
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter
-- Cache files will be written to: /home/xxx/.cache/zephyr
-- Zephyr version: 3.6.99 (/home/xxx/RTOS-exploration/zephyrproject/zephyr)
CMake Error at /home/xxx/RTOS-exploration/zephyrproject/zephyr/cmake/modules/boards.cmake:193 (message):
  Error finding board: nucleo_h563zi

  Error message: Traceback (most recent call last):

    File "/home/xxx/RTOS-exploration/zephyrproject/zephyr/scripts/list_boards.py", line 11, in <module>
      import pykwalify.core

  ModuleNotFoundError: No module named 'pykwalify'

Call Stack (most recent call first):
  /home/xxx/RTOS-exploration/zephyrproject/zephyr/cmake/modules/zephyr_default.cmake:132 (include)
  /home/xxx/RTOS-exploration/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/xxx/RTOS-exploration/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:159 (include_boilerplate)
  CMakeLists.txt:6 (find_package)


-- Configuring incomplete, errors occurred!
```

**Comment by sebistark** [comment, marked as answer]
I found a workaround.
put the following two commands in CMakeLists before "findpackage()"

```
set(Python3_EXECUTABLE "/home/xxx/RTOS-exploration/zephyrproject/.venv/bin/activate")
set(PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
```

This path is were python venv is located.

**Comment by yashi** [comment]
Setting some variable in `CMakeLists.txt` might work. But but the standard way to use venv is
```
$ source <venv>/bin/activate
```
See https://docs.python.org/3/library/venv.html#how-venvs-work

Or you can use one of virtual environment tools, such as conda, pipenv, poetry.
Assuming you have already installed all python dependencies:

```
❯ cd zephyr-project
❯ pipenv shell  # or `source .venv/bin/activate`
❯ export ZEPHYR_BASE=`pwd`/zephyr
❯ export BOARD=native_sim
❯ cmake -B builddir -S zephyr/samples/hello_world
❯ ninja -C builddir
```

**Reply by sebistark** [reply]
Thank you for your reply.

I think you are right. This is the proper way, if you do everything in the shell.
However, I wanted to find a solution, where one can import CMakeLists.txt in an IDE (e.g. Eclipse or STMCubeIDE) and compile it from there.

**Reply by yashi** [reply]
OK. I can understand that.

However, all I did was set a few environment variables.

What `source .venv/bin/activate` does is set environment variables. It basically sets:
- `PATH`
- `VIRTUAL_ENV`
- `VIRTUAL_ENV_PROMPT`

These are called environment variable because they might differ in other situations. If the `CMakeLists.txt` is used only by you, your solution should be fine. But if you are planning to work with your team, it's gonna be a problem.

ps.
VS code seems to have venv support: https://code.visualstudio.com/docs/python/environments

I'm not an IDE guy but Eclipse and STMCubeIDE should have some support for Python as well.

Anyway, happy hacking :smile:

**Reply by sebistark** [reply]
Thanks a lot 😄
